### PR TITLE
Migrate bls crates to blsful

### PR DESCRIPTION
**Pull request by shawn-zil**
Description in the linked ticket.

### CODE REVIEW

**Comment by JamesHinshelwood** [comment]
Performance of processing blocks looks miles better with this PR, nice one :) A `-59.179%` change in processing time per block and `+144.97%` improvement in throughput.

process-blocks-both

(Blame criterion for generating a misleading graph where the axes don't start at zero :stuck_out_tongue:)

File: zilliqa/src/crypto.rs [hunk 1]
```diff
@@ -20,20 +22,44 @@ use sha3::{Digest, Keccak256};
...
+        NodeSignature(Signature::<Bls12381G2Impl>::Basic(
+            blsful::inner_types::G2Projective::identity(),
+        ))
     }
 
     pub fn from_bytes(bytes: &[u8]) -> Result<NodeSignature> {
-        Ok(NodeSignature(bls_signatures::Signature::from_bytes(bytes)?))
+        // Default to Basic signatures - it's the normal signature.
+        Ok(NodeSignature(Signature::<Bls12381G2Impl>::Basic(
+            blsful::inner_types::G2Projective::from_compressed(bytes.try_into()?)
```
Comment (JamesHinshelwood) [inline zilliqa/src/crypto.rs:38]: Just to double check - You verified that `bls_signatures::Signature::from_bytes` was treating the bytes as a compressed signature?

File: zilliqa/src/crypto.rs [hunk 2]
```diff
@@ -20,20 +22,44 @@ use sha3::{Digest, Keccak256};
...
+                .expect("blst_p2_uncompress() error"),
```
Comment (JamesHinshelwood) [inline zilliqa/src/crypto.rs:39]: Should we propagate the error instead? `from_bytes` is returning a `Result`, but this implementation never returns an `Err`. It seems like `from_compressed` will fail if `!p.is_on_curve()` or `!p.is_torsion_free()` and potentially for more reasons if the underlying C library doesn't return success.

File: zilliqa/src/crypto.rs [hunk 3]
```diff
@@ -20,20 +22,44 @@ use sha3::{Digest, Keccak256};
...
-        let signatures: Vec<_> = signatures.iter().map(|s| s.0).collect();
-        Ok(NodeSignature(bls_signatures::aggregate(&signatures)?))
+        let signatures = signatures.iter().map(|s| s.0).collect_vec();
+
+        // IETF standards say N >= 1
+        // Handles single case where N == 1, as AggregateSignature::from_signatures() only handles N > 1.
+        if signatures.len() < 2 {
+            let g = blsful::inner_types::G2Projective::identity();
+            return Ok(NodeSignature(match signatures[0] {
+                Signature::Basic(s) => Signature::Basic(g + s),
```
Comment (JamesHinshelwood) [inline zilliqa/src/crypto.rs:51]: Surely `identity + s = s`? Why the extra step?

File: zilliqa/src/crypto.rs [hunk 4]
```diff
@@ -20,20 +22,44 @@ use sha3::{Digest, Keccak256};
...
+        )))
     }
 
     pub fn aggregate(signatures: &[NodeSignature]) -> Result<NodeSignature> {
-        let signatures: Vec<_> = signatures.iter().map(|s| s.0).collect();
-        Ok(NodeSignature(bls_signatures::aggregate(&signatures)?))
+        let signatures = signatures.iter().map(|s| s.0).collect_vec();
+
+        // IETF standards say N >= 1
+        // Handles single case where N == 1, as AggregateSignature::from_signatures() only handles N > 1.
```
Comment (JamesHinshelwood) [inline zilliqa/src/crypto.rs:47]: https://github.com/hyperledger-labs/agora-blsful/issues/10 - Feel free to reference here in case it gets fixed.

**Review by JamesHinshelwood — DISMISSED** [review]
LGTM

File: zilliqa/src/crypto.rs (hunk 3 above)
Comment (shawn-zil) [reply zilliqa/src/crypto.rs:51]: I followed bls-signatures, which used a fold() with the offset of g. But you're right, it produces the same outcome. I'll remove the extra step.

File: zilliqa/src/crypto.rs (hunk 1 above)
Comment (shawn-zil) [reply zilliqa/src/crypto.rs:38]: Yes, there's a check in `bls_signatures::Signature::from_bytes`:
```
    if raw.len() != G2_COMPRESSED_SIZE {
        return Err(Error::SizeMismatch);
    }
```

Comment (shawn-zil) [reply zilliqa/src/crypto.rs:38]: But I don't seem to find a length check in the underlying library, so I've added length checks in ours.

**Review by JamesHinshelwood — APPROVED** [review]

**Pull request merged**